### PR TITLE
Create crud user

### PR DESCRIPTION
**Pull request by jolivo07**

### CODE REVIEW

File: app/actions/users.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+	"fmt"
+	"mjm/app/models"
+	"net/http"
+
+	"github.com/gobuffalo/buffalo"
+	"github.com/gobuffalo/buffalo/render"
+	"github.com/gobuffalo/pop/v6"
+)
+
+func List(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: This name is too general when we are working with several models or several CRUDs, this should be called **ListUsers**, in this way there could be several actions to list example ListDepartment, etc.

File: app/actions/users.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		return err
+	}
+
+	c.Set("pagination", q.Paginator)
+	c.Set("users", users)
+
+	return c.Render(http.StatusOK, r.HTML("/user/index.plush.html"))
+}
+
+func Show(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

File: app/actions/users.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	if err := tx.Find(user, c.Param("user_id")); err != nil {
+		return c.Error(http.StatusNotFound, err)
+	}
+
+	c.Set("user", user)
+
+	return c.Render(http.StatusOK, r.HTML("/user/show.plush.html"))
+}
+
+func New(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

File: app/actions/users.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+		return err
+	}
+
+	c.Set("departments", departments.Map())
+	c.Set("user", &models.User{})
+
+	return c.Render(http.StatusOK, r.HTML("/user/new.plush.html"))
+}
+
+func Create(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

File: app/actions/users.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		c.Set("user", user)
+		return c.Render(http.StatusUnprocessableEntity, r.HTML("/user/new.plush.html"))
+	}
+
+	c.Flash().Add("success", "user.created.success")
+
+	return c.Redirect(http.StatusSeeOther, "usersPath()", render.Data{"user_id": user.ID})
+}
+
+func Edit(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

File: app/actions/users.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	if err := tx.All(&departments); err != nil {
+		return err
+	}
+	c.Set("departments", departments.Map())
+	c.Set("user", user)
+
+	return c.Render(http.StatusOK, r.HTML("/user/edit.plush.html"))
+}
+
+func Update(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

File: app/actions/users.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+		c.Set("user", user)
+		return c.Render(http.StatusUnprocessableEntity, r.HTML("/user/edit.plush.html"))
+	}
+
+	c.Flash().Add("success", "user.updated.success")
+
+	return c.Redirect(http.StatusSeeOther, "usersPath()", render.Data{"user_id": user.ID})
+}
+
+func Destroy(c buffalo.Context) error {
```
Comment (mperezlamadrid) [inline app/actions/users.go]: Check the first comment

**Review by mperezlamadrid — APPROVED** [review]

File: app/actions/user/users_test.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+	res := as.HTML("/create-user").Post(user)
+
+	as.Equal(res.Code, http.StatusSeeOther)
+	as.Equal("/users/", res.Location())
+	userArray := []models.User{}
+	as.DB.All(userArray)
+	for _, v := range userArray {
+		as.Equal(v.FirstName, user.FirstName)
+	}
```
Comment (rdiazrovira) [inline app/actions/user/users_test.go]: Please add a validation of the length of "userArray", and I suggest you rename the variable. Just "users" would be ok.

File: app/actions/user/users_test.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+	userUpdate.FirstName = "Joaquin"
+	userUpdate.LastName = "asfasf"
+	userUpdate.ID = user.ID
+	userUpdate.DepartmentID = department.ID
+
+	res := as.HTML("/delete-user/" + user.ID.String()).Delete()
+	as.Equal(res.Code, http.StatusSeeOther)
+	as.Equal("/users/", res.Location())
+
+}
```
Comment (rdiazrovira) [inline app/actions/user/users_test.go]: Can you validate if the user has actually been deleted?

File: app/actions/user/users_test.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+	res := as.HTML("/new-users/").Get()
+	as.Equal(http.StatusOK, res.Code)
+	body := res.Body.String()
+	as.Contains(body, user.FirstName)
+	as.Contains(body, user.LastName)
+	as.Contains(body, user.PhoneNumber)
+	as.Contains(body, user.DNI)
+	as.Contains(body, "New User")
+
+}
```
Comment (rdiazrovira) [inline app/actions/user/users_test.go]: What is the purpose of this test?

**Review by rdiazrovira — CHANGES_REQUESTED** [review]
@jolivo07 Take a look at my comments here.

File: app/actions/user/users_test.go [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+	}
+
+	as.NoError(as.DB.Create(user))
+
+	res := as.HTML("/users/").Get()
+	as.Equal(http.StatusOK, res.Code)
+	body := res.Body.String()
+	as.Contains(body, user.FirstName)
+
+}
```
Comment (rdiazrovira) [inline app/actions/user/users_test.go]: It looks like you are testing the handler "List", not "Edit", can you review this?

Comment (jolivo07) [reply app/actions/user/users_test.go]: i'll get on that

**Review by rdiazrovira — APPROVED** [review]

**Pull request merged**